### PR TITLE
Support to DNS, IP and Email SAN fields

### PR DESCRIPTION
**Pull request by seocam**

### CODE REVIEW

**Comment by seocam** [comment]
This PR depends on #3 but the commit `Support to DNS, IP and Email SAN fields` can be reviewed.

File: examples/basic_self_signed.yml [hunk 1]
```diff
@@ -9,4 +9,4 @@
         ca: self-sign
 
   roles:
-    - certificate
+    - linux-system-roles.certificate
```
Comment (richm) [inline examples/basic_self_signed.yml:12]: This change looks unrelated to the subject of the PR, unless it is necessary to run the test

File: module_utils/certificate/providers/base.py [hunk 2]
```diff
@@ -36,10 +37,14 @@ def load_from_params(cls, module, params):
         # pylint: disable=protected-access
         cert_like = cls(module)
 
-        map_attrs = ["dns", "ip"]
+        map_attrs = ["dns", "ip", "email"]
         info = {k: v for k, v in params.items() if k in map_attrs}
 
         info["common_name"] = cert_like._get_common_name_from_params(params)
+        if info.get("ip"):
```
Comment (richm) [inline module_utils/certificate/providers/base.py:44]: this works, but the typical python idiom is to use `if "ip" in info` since you are not otherwise using the result of `get()`

File: module_utils/certificate/providers/certmonger.py [hunk 3]
```diff
@@ -136,6 +136,18 @@ def request_certificate(self):
         # Set Domains
         for dns in self.csr.dns:
             command += ["-D", dns]
+        if not self.csr.dns:
+            command += ["-D", ""]
```
Comment (richm) [inline module_utils/certificate/providers/certmonger.py:140]: this seems odd - why wouldn't you just omit the `-D` argument?  Or does that have a different semantic meaning than `-D ""`?

File: tests/tests_wrong_provider.yml [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+      block:
+        - import_role:
+            name: linux-system-roles.certificate
+        - fail:
+            msg: "Certificate issued with nonexistent provider 'fake-provider'."
+      rescue:
+        - name:
+          assert:
+            that:
+              - ansible_failed_task.name == "Ensure certificate requests"
```
Comment (richm) [inline tests/tests_wrong_provider.yml:26]: `ansible_failed_task` is currently broken - https://github.com/ansible/ansible/issues/57399

See https://github.com/linux-system-roles/logging/blob/master/tests/tests_files_elasticsearch_certs_incomplete.yml#L24 for the current best practices idiom of testing for a failure with `block/rescue` by using a dummy `name: unreachable task` and `ansible_failed_result`

File: examples/basic_self_signed.yml (hunk 1 above)
Comment (seocam) [reply examples/basic_self_signed.yml:12]: This is actually part of PR #3 but it's also not required by that PR itself. It's there because running the tests I'm running the examples manually to ensure they work.

File: module_utils/certificate/providers/certmonger.py (hunk 3 above)
Comment (seocam) [reply module_utils/certificate/providers/certmonger.py:140]: It's indeed weird but it's required. 

For example, if we have 2 runs with the following settings:
First:
```
    certificate_requests:                                                                                 
      - name: mycert                                                                                      
        dns:                                                                                              
          - www.example.com                                                                                                                                    
        ca: self-sign    
```
Second:
```                                                                                
    certificate_requests:                                                                                 
      - name: mycert                                                                                      
        ip:                                                                                               
          - 192.0.2.12                                                                                         
        ca: self-sign    
```

If I don't set `-D ""` certmonger will leave `www.example.com` in the certificate which I don't think it's the behavior we want the role to have.

File: module_utils/certificate/providers/base.py (hunk 2 above)
Comment (seocam) [reply module_utils/certificate/providers/base.py:44]: I actually want to get the value and verify that's not `None` neither `[]`.

Comment (richm) [reply module_utils/certificate/providers/base.py:44]: ah, ok - then what you have is shorter and more readable then `if "ip" in info and info["ip"]`

Comment (seocam) [reply module_utils/certificate/providers/base.py:44]: @richm alright. Although actually a little longer the intention is obviously a lot clearer in that form. I'll change.

Comment (richm) [reply module_utils/certificate/providers/base.py:44]: I think you should leave it as `if info.get("ip"):` - I think that is shorter and more readable than the alternative (imo)

Comment (seocam) [reply module_utils/certificate/providers/base.py:44]: I think that if I had `if "ip" in info and info["ip"]` you probably wouldn't have the doubt you had. That's why I assumed the intention would be clearer in the longer form.
But this is so small that I'm ok with either way.

File: tests/tests_wrong_provider.yml (hunk 4 above)
Comment (seocam) [reply tests/tests_wrong_provider.yml]: Updated in b8a4c36c7b30d2da633b06937ce862c674cc0c91

File: module_utils/certificate/providers/certmonger.py [hunk 5]
```diff
@@ -110,6 +110,7 @@ def exists_in_certmonger(self):
 
     def request_certificate(self):
         """Issue or update a certificate using certmonger."""
+        # pylint: disable=useless-else-on-loop
```
Comment (richm) [inline module_utils/certificate/providers/certmonger.py:113]: Why is the `else` useless?  Does the linter think that e.g. `self.csr.dns` will always have at least 1 element?

Comment (seocam) [reply module_utils/certificate/providers/certmonger.py:113]: `pylint` complains that there is an `else` in a `for` loop when there is no `break` inside the loop. The `for-else` was designed for fallback in cases where the programmer is looking for something in the loop and can't find it.
Since we are using it in a different way we need to disable this warning.

Initially I've used `if not` to avoid the lint complains without having to explicitly silent it.

Comment (richm) [reply module_utils/certificate/providers/certmonger.py:113]: Ok - I see - `for/else` is supposed to be used with a `break` - I have typically used it without a break, but not in linted code . . . then I'll leave it up to you if you want to revert back to the `if not xxx` version, or keep the current code with the pylint suppression.

Comment (seocam) [reply module_utils/certificate/providers/certmonger.py:113]: Ok. I fine with the current way.

**Comment by richm** [comment]
lgtm

**Comment by pcahyna** [comment]
I believe you should omit all those `become: yes` in tests, I don't think we use them elsewhere. In general, our roles do not make much sense to execute as other user than root or without become.

**Comment by amore17** [comment]
> I believe you should omit all those `become: yes` in tests, I don't think we use them elsewhere. In general, our roles do not make much sense to execute as other user than root or without become.
IMO we can keep `become: yes` as it is as these tests will be running in downstream IDM-CI also.
